### DISCUSSION TITLE
Possibility of add constraint on predictions

### DISCUSSION THREAD

**Discussion by yewalenikhil65** [DENG-MIT/CRNN#19] [Ideas]
@jiweiqi It would be nice to add previously known physical constraint on the `pred = ODEProblem(CRNN)`.
Example, in PBE it's well known that conservation law is followed.. so if we could add this constraint in loss function , i have a hunch that it might reduce the training time.

Same thoughts applies to robertson's case, where the third species is a combination of first and second species

**Comment by jiweiqi** [comment]
Agree. What specific regularization do you want to impose?

For combustion, I am using elemental conservation, which can be achieved by projecting the stoichiometric coefficients into the elemental conservation space. If you are interested, I can point out the code that I have written for it.

The issue with adding regularization into the loss function is that it could make the optimization more difficult and thus degrade the benefit. So it is heuristic, it is an option to try but not necessarily give better performance.

**Reply by yewalenikhil65** [reply]
> Agree. What specific regularization do you want to impose?
> 

example in robers case, we know that y1+y2+y3=constant(from initial condition) 

> For combustion, I am using elemental conservation, which can be achieved by projecting the stoichiometric coefficients into the elemental conservation space. If you are interested, I can point out the code that I have written for it.
>
Yes I would like to know about this
> The issue with adding regularization into the loss function is that it could make the optimization more difficult and thus degrade the benefit. So it is heuristic, it is an option to try but not necessarily give better performance.
>
Well, there is a tutorial based on this on DiffEqFlux where they add conservation structure within the Neural-networks definition.. Maybe we can do that? https://diffeqflux.sciml.ai/stable/examples/physical_constraints/

**Reply by jiweiqi** [reply]
Oh, sure, we can add this constraint, but there is a lot of open questions about this constrain (which converts the problem to a DAE problem). So this is case by case, if something not works, this can be an option.

**Reply by yewalenikhil65** [reply]
> Oh, sure, we can add this constraint, but there is a lot of open questions about this constrain (which converts the problem to a DAE problem). So this is case by case, if something not works, this can be an option.
>
Ours isn't DAE way, rather CRNN apporach uses `ODEProblem` or `NeuralODE` over `CRNN(u,p,t)` which doesnt have facility for physical constraints. 
any way to directly specify the constraint while defining `CRNN` itself ? eg: `y1+y2+y3 = constant` means `dy1/dt+ dy2/dt + dy3/dt =0`. this could be a constraint we adapt for the output of `CRNN` itself ? i am not clear how to impose this in code.

one way i can think of is defining our own callback that should be used as argument whenever we call `solve(predict_ode(CRNN))` function..https://diffeq.sciml.ai/dev/features/callback_library/
There is an example of `ManifoldProjection` that uses `energy conservation` for mechanical problems.. That could be useful

**Reply by jiweiqi** [reply]
For Robertson's problem. I don't think it is necessary to add this constrain since it is already informed by the data. Such regularization is more useful when the problem is ill-posed, for example, x3 can be measured and we can use this conservation law to approximate x3.

**Reply by yewalenikhil65** [reply]
> For Robertson's problem. I don't think it is necessary to add this constrain since it is already informed by the data. Such regularization is more useful when the problem is ill-posed, for example, x3 can be measured and we can use this conservation law to approximate x3.
>
you mean when one of the species data is not available ?

**Reply by jiweiqi** [reply]
Yes, some species data are not available
